modules/annotations/shared/src/main/scala/io/circe/derivation/annotations/Configuration.scala

```diff
@@ -19,32 +20,52 @@ sealed trait Configuration {
    */
   def transformMemberNames: String => String
 
+  def useDefaults: Boolean
+
+  def discriminator: Discriminator
+
   protected def getA(transformMemberNames: String => String): Config
 
+  protected def applyDiscriminator(discriminator: Discriminator): Config
+
   /** Creates a configuration which produces snake cased member names */
   final def withSnakeCaseMemberNames: Config =
     getA(renaming.snakeCase)
 
   /** Creates a configuration which produces kebab cased member names */
   final def withKebabCaseMemberNames: Config =
     getA(renaming.kebabCase)
+
+  final def withDiscriminatorName(name: String): Config =
+    applyDiscriminator(Discriminator.Embedded(name))
+
+  final def withTypeDiscriminator: Config =
+    applyDiscriminator(Discriminator.TypeDiscriminator)
 }
 
 object Configuration {
+  import io.circe.derivation.Discriminator
 
   /** Configuration allowing customisation of JSON produced when encoding or
    *  decoding.
    *
    *  This configuration creates *both* encoder and decoder.
    */
   final case class Codec(
-    transformMemberNames: String => String
+    transformMemberNames: String => String,
```

Comment:
Now that we support ADTs we should probably have `transformConstructorNames` as well, but I can add that in a follow-up.